src/codecs/decoders.ts

```diff
@@ -0,0 +1,58 @@
+import * as wasmWebP from './webp/decoder';
+import * as browserWebP from './browser-webp/decoder';
+
+import { createImageBitmapPolyfill, sniffMimeType } from '../lib/util';
+
+export interface Decoder {
+  name: string;
+  decode(file: File): Promise<ImageBitmap>;
+  isSupported(): Promise<boolean>;
+  canHandleMimeType(mimeType: string): boolean;
+}
+
+// We load all decoders and filter out the unsupported ones.
+export const decodersPromise: Promise<Decoder[]> = Promise.all(
+  [
+    browserWebP,
+    wasmWebP,
+  ]
+  .map(async (decoder) => {
+    if (await decoder.isSupported()) {
+      return decoder;
+    }
+    return null;
+  }),
+// TypeScript is not smart enough to realized that I’m filtering all the falsy
+// values here.
+).then(list => list.filter(item => !!item)) as any as Promise<Decoder[]>;
+
+export async function findDecoder(file: File): Promise<Decoder | undefined> {
+  const decoders = await decodersPromise;
+  const mimeType = await sniffMimeType(file);
+  if (!mimeType) {
+    return;
+  }
+  return decoders.find(decoder => decoder.canHandleMimeType(mimeType));
+}
+
+const nativelySupportedMimeTypes = [
+  'image/jpeg',
+  'image/png',
+  'image/gif',
+];
+
+export async function decodeFile(file: File): Promise<ImageBitmap> {
+  const mimeType = await sniffMimeType(file);
+  if (!mimeType) {
+    throw new Error('Could not determine mime type');
+  }
+  if (nativelySupportedMimeTypes.includes(mimeType)) {
```

Comment:
I still think we're doing this backwards. Instead of whitelisting supported types, we should have a list of types we have wasm decoders for. If the incoming image is one of those types, throw it at that decoder (which may just throw it at the browser's decoder), otherwise, throw it at the browser's decoder.

That means we can support types only supported by particular browsers without effort. Right now this PR prevents loading jpeg2000 even though at least Safari supports that format.

Happy to take a stab at it if I'm explaining badly.